packages/expo-router/src/layouts/ModalStack.web.tsx

```diff
@@ -0,0 +1,364 @@
+'use client';
+import {
+  createNavigatorFactory,
+  ParamListBase,
+  StackNavigationState,
+  StackRouter,
+  useNavigationBuilder,
+  useTheme,
+} from '@react-navigation/native';
+import { NativeStackView } from '@react-navigation/native-stack';
+import React from 'react';
+import { Drawer } from 'vaul';
+
+import { ExtendedStackNavigationOptions } from './StackClient';
+import { withLayoutContext } from './withLayoutContext';
+import modalStyles from '../../assets/modal.module.css';
+
+type Props = {
+  initialRouteName?: string;
+  screenOptions?: ExtendedStackNavigationOptions;
+  children: React.ReactNode;
+};
+
+type CSSWithVars = React.CSSProperties & {
+  [key: `--${string}`]: string | number;
+};
+
+// Helper to determine if a given screen should be treated as a modal-type presentation
+function isModalPresentation(
+  options?: Partial<Pick<ExtendedStackNavigationOptions, 'presentation'>> | null
+) {
+  const presentation = options?.presentation;
+  return (
+    presentation === 'modal' ||
+    presentation === 'formSheet' ||
+    presentation === 'fullScreenModal' ||
+    presentation === 'containedModal'
+  );
+}
+
+function ModalStackNavigator({ initialRouteName, children, screenOptions }: Props) {
+  const { state, navigation, descriptors, NavigationContent, describe } = useNavigationBuilder(
+    StackRouter,
+    {
+      children,
+      screenOptions,
+      initialRouteName,
+    }
+  );
+
+  return (
+    <NavigationContent>
+      <ModalStackView
+        state={state}
+        navigation={navigation}
+        descriptors={descriptors}
+        describe={describe}
+      />
+    </NavigationContent>
+  );
+}
+
+function ModalStackView({
+  state,
+  navigation,
+  descriptors,
+  describe,
+}: {
+  state: StackNavigationState<ParamListBase>;
+  navigation: any;
+  descriptors: Record<
+    string,
+    {
+      navigation: any;
+      route: any;
+      options: ExtendedStackNavigationOptions;
+      render: () => React.ReactNode;
+    }
+  >;
+  describe: any;
+}) {
```

Comment:
And this can be replaced with

```ts
type ModalStackViewProps = Omit<
  ReturnType<
    typeof useNavigationBuilder<
      StackNavigationState<ParamListBase>,
      StackRouterOptions,
      StackActionHelpers<ParamListBase>,
      NativeStackNavigationOptions,
      NativeStackNavigationEventMap
    >
  >,
  'NavigationContent'
>;
```